diff --git a/cgroup.cc b/cgroup.cc
index a72e35f..15dbc52 100644
--- a/cgroup.cc
+++ b/cgroup.cc
@@ -136,10 +136,10 @@ static bool initNsFromParentCpu(nsjconf_t* nsjconf, pid_t pid) {
 
 	std::string cpu_ms_per_sec_str = std::to_string(nsjconf->cgroup_cpu_ms_per_sec * 1000U);
 	RETURN_ON_FAILURE(
-	    writeToCgroup(cpu_cgroup_path + "/cpu.cfs_quota_us", cpu_ms_per_sec_str, "cpu quota"));
+	    writeToCgroup(cpu_cgroup_path + "/cpu.cfs_period_us", "1000000", "cpu period"));
 
 	RETURN_ON_FAILURE(
-	    writeToCgroup(cpu_cgroup_path + "/cpu.cfs_period_us", "1000000", "cpu period"));
+	    writeToCgroup(cpu_cgroup_path + "/cpu.cfs_quota_us", cpu_ms_per_sec_str, "cpu quota"));
 
 	return addPidToTaskList(cpu_cgroup_path, pid);
 }
